[mlir][vector] Propagate alignment from vector to llvm dialects.

amd-eochoalo opened the pull request:
Allows alignment to be propagated correctly from vector to LLVM dialect operations.

amd-eochoalo committed:
Pass alignment from vector.compressstore to llvm.compressstore

amd-eochoalo committed:
Pass alignment from vector.expandload to llvm.intr.masked.expandload

amd-eochoalo committed:
Pass alignment from vector.scatter to llvm.intr.masked.scatter

amd-eochoalo committed:
Pass alignment from vector.gather to llvm.intr.masked.gather

krzysz00 reviewed (approved):
LGTM

banach-space commented on mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp:
```diff
@@ -299,8 +299,9 @@ class VectorGatherOpConversion
     // Resolve alignment.
-    unsigned align;
-    if (failed(getVectorToLLVMAlignment(*this->getTypeConverter(), vType,
+    unsigned align = gather.getAlignment().value_or(0);
+    if (!align &&
+        failed(getVectorToLLVMAlignment(*this->getTypeConverter(), vType,
```
Note that this means that explicit alignment takes priority over e.g. `--convert-vector-to-llvm='use-vector-alignment=1`. This feels like the right design decision, but we should make sure that it is documented (perhaps just add a comment here?) and tested (e.g. in "use-vector-alignment.mlir").

Similar comment for scatter. Thanks!

banach-space replied:
Thanks! Sorry I didn't replay earlier, I was on a sick leave.

amd-eochoalo committed:
Add tests

amd-eochoalo committed:
[mlir] Document priority of alignment options.

Alignment can be specified either explicitly via
attributes in vector operations or via the option
to use-vector-alignment=<N> in the --convert-vector-to-llvm
pass. The explicit attribute takes precedent over the option
used as input to the pass.

amd-eochoalo committed:
[mlir] Document alignment priority in option description.

dcaballe reviewed (approved):
LGTM, thanks!

amd-eochoalo committed:
Merge branch 'main' into eochoa/2025-08-13/alignment-flow
